BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
-      file_group_id ||= current_project.try(:id) || current_user.own_group.id
-      if file_group_id == Group.everyone.id ||          # We won't register a file in group 'everyone'
-         ! current_user.available_groups.where(:id => file_group_id).exists? # or that is not in one of our groups
-         file_group_id = current_user.own_group.id
-      end
+          when :copy
+            new = userfile.provider_copy_to_otherprovider(target_dp)
+            userfile.delete rescue true # Not destroy(), as the contents must be kept.
+            userfile.destroy_log rescue true
+            (userfile = new).set_size!
+          end
 
-      subclass = Class.const_get(subtype)
-      userfile = subclass.new( :name             => basename,
-                               :user_id          => @as_user.id, # cannot use current_user, since it might be a vault_ssh dp
-                               :group_id         => file_group_id,
-                               :data_provider_id => @provider.id )
-
-      registered_file = registered_files[basename]
-      if registered_file
-        previously_registered_userfiles << registered_file
-      elsif userfile.save
-        newly_registered_userfiles << userfile
-        userfile.addlog_context(self, "Registered on DataProvider '#{@provider.name}' as '#{userfile.name}' by #{current_user.login}.")
-      else
-        flash[:error] += "Error: could not register #{subtype} '#{basename}'... maybe the file exists already?\n"
-        num_skipped += 1
+          userfile.set_size!
+          succeeded << userfile
+        rescue => e
+          (failed[e.message] ||= []) << userfile
+        end
       end
 
-    end # loop to register/unregister files
+      mangled_action = (post_action == :move ? 'mov' : 'copy')
+      generic_notice_messages(mangled_action, succeeded, failed)
+    end
+
+    # Generate a complete response matching the old API
+    flash[:notice] += "Registering #{userfiles_count} userfile(s) in background.\n"
+    api_response = generate_register_response.merge({
+      :newly_registered_userfiles      => registered,
+      :previously_registered_userfiles => already_registered
+    })
 
-    if num_skipped > 0
-      flash[:notice] += "Skipped #{num_skipped} files.\n"
+    respond_to do |format|
+      format.html { redirect_to :action => :browse }
+      format.json { render :json => api_response }
+      format.json { render :xml  => api_response }
     end
+  end
 
-    if newly_registered_userfiles.size > 0
-      clear_browse_provider_local_cache_file(@as_user, @provider) unless request.format.to_sym == :xml || request.format.to_sym == :json
-      flash[:notice] += "Registered #{newly_registered_userfiles.size} files.\n"
-      if @as_user != current_user
-        flash[:notice] += "Important note! Since you were browsing as user '#{@as_user.login}', the files were registered as belonging to that user instead of you!\n"
+  # Unregister a list of files from CBRAIN
```

Comment:
Still, I'd like before each action maybe 3 lines of comment instead of just one. Just to explain quickly what the action does, since there are subtleties.